---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  eval = FALSE,
  collapse = TRUE,
  comment = "#>",
   dpi=900,
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```


# Chevreul

This package includes a set of Shiny apps for exploring single cell RNA datasets processed with <a href="https://github.com/satijalab/seurat" target="_blank" rel="noopener noreferrer">Seurat</a>

A demo using a pancreas dataset from the Seurat team is available <a href="http://cobriniklab.saban-chla.usc.edu:3838/chevreul_demo/" target="_blank" rel="noopener noreferrer">here</a>

There are also convenient functions for:

-   Clustering and Dimensional Reduction of Raw Sequencing Data.
-   <a href="https://satijalab.org/seurat/v3.0/pancreas_integration_label_transfer.html" target="_blank" rel="noopener noreferrer">Integration and Label Transfer</a>
-   Louvain Clustering at a Range of Resolutions
-   Cell cycle state regression and labeling
-   RNA velocity calculation with <a href="https://velocyto.org/" target="_blank" rel="noopener noreferrer">Velocyto.R</a> and <a href="https://scvelo.readthedocs.io/" target="_blank" rel="noopener noreferrer">scvelo</a>

## Installation

You can install the released version of chevreul from <a href="https://github.com/whtns/chevreul" target="_blank" rel="noopener noreferrer">github</a> with:

### Install locally and run in three steps:

```{r}
install.packages("devtools")
devtools::install_github("whtns/chevreul")
chevreul::create_project_db()
```

### Install locally (custom location!) and run in three steps:

```{r}
devtools::install_github("whtns/chevreul")
chevreul::create_project_db(destdir='/your/path/to/app')
```

## Site

You can view documentation on the <a href="https://whtns.github.io/chevreul" target="_blank" rel="noopener noreferrer">chevreul website</a>

## How To

## Getting Started

```{r}
library(chevreul)
library(Seurat)
library(tidyverse)
library(ggraph)
```

## TLDR

chevreul provides a single command to:

-   construct a Seurat object

-   filter genes by minimum expression and ubiquity

-   normalize and scale expression by any of several methods packaged in Seurat

## Run clustering on a single seurat object

By default clustering will be run at ten different resolutions between 0.2 and 2.0. Any resolution can be specified by providing the resolution argument as a numeric vector.

```{r, cache = TRUE, eval = FALSE}
clustered_seu <- clustering_workflow(panc8, experiment_name = "seurat_pancreas", organism = "human")
```

## Get a first look at a processed dataset using an interactive shiny app

```{r, eval = FALSE}
minimalSeuratApp(clustered_seu)
```

## Set up a seurat object

We start with a gene by cell matrix of count/UMI values and a table of cell metadata

```{r, eval=FALSE}
panc8_count_mat[1:5, 1:5]

head(panc8_metadata)
```

```{r, echo=FALSE}
knitr::kable(panc8_count_mat[1:5, 1:5])

knitr::kable(head(panc8_metadata))
```

we can create a seurat object in the usual manner

```{r}
myseu <- CreateSeuratObject(panc8_count_mat, assay = "gene", meta.data = panc8_metadata)
myseu
```

## Preprocess the seurat object

Chevreul includes a handy function to preprocess the data that handles normalization and scaling required for downstream analysis. Preprocessing is performed using existing Seurat functions. If needed, parameters can be specified by the user.

```{r, message = FALSE}
myseu <- seurat_preprocess(myseu)
```

This single function includes sub-functions that normalizes, identifies highly variable features and scales the data:

-   The normalizing sub-function performs log normalization using a default scale factor of 10,000.

```{r}
preprocess_seu <- NormalizeData(myseu, verbose = FALSE)
```

-   After normalization, subset of features that exhibit high cell-to-cell variation in the dataset are identified. By default, 2,000 features per dataset are returned by this function.

```{r}
preprocess_seu <- FindVariableFeatures(preprocess_seu, selection.method = "vst", 
    verbose = FALSE)
```

-   Finally, the data is scaled by applying linear transformation. This step shifts the gene expression, so that the mean expression across cells is 0 and scales the gene expression, so that the variance across cells is 1.

```{r, message = FALSE}
pre_process_seu <- ScaleData(preprocess_seu)
```

## Perform dimension reduction

Chevreul also implements a standardized dimension reduction step to select variable features at a user-specified threshold and perform PCA, tSNE, and UMAP. The default assay the dimension reduction is being run on is "gene".

```{r, results=FALSE, warning = FALSE, message=FALSE}
myseu <- seurat_reduce_dimensions(myseu, assay = "RNA") 
```

This function includes existing seurat functions which performs dimension reduction techniques.

-   Perform PCA: Runs a PCA dimensionality reduction.

```{r, results=FALSE, warning = FALSE, message=FALSE}
Dim_Red_seu <- RunPCA(myseu,features = VariableFeatures(myseu), 
      do.print = FALSE)
```

-   Perform tSNE: Runs t-SNE dimensionality reduction on selected features

```{r, results=FALSE, warning = FALSE, message=FALSE}
Dim_Red_seu <- RunTSNE(Dim_Red_seu, dims = 1:30)
DimPlot(Dim_Red_seu, reduction = "tsne")

```

-   Perform UMAP: Runs the Uniform Manifold Approximation and Projection (UMAP) dimensional reduction technique.

```{r, results=FALSE, warning = FALSE, message=FALSE, fig.width=7, fig.height=5}
Dim_Red_seu <- RunUMAP(Dim_Red_seu, dims = 1:30)
DimPlot(Dim_Red_seu, reduction = "umap")
```

## Community detection by clustering

Clustering analysis is performed via Louvain(default) or alternative algorithms available in Seurat. Clustering is performed at a range of resolutions with default value ranging from 0.2 to 2 and pca reduction

```{r, message = FALSE, result = FALSE, chache = FALSE, warning = FALSE}
 seu <- seurat_cluster(seu = Dim_Red_seu, resolution = seq(0.2, 2, by = 0.2) )
```

This function produces clustering analysis via two steps performed using two different sub-functions

-   `FindNeighbours`: This function computes the nearest neighbors for a given dataset using k-nearest neighbor algorithm.

-   `FindClusters`: The output from FindNeighbours is then used to compute 

## Split included dataset based on collection technology

Chevreul includes a function, `SplitObject`, which is capable of splitting the dataset into subsets based on a single attribute indicated by the split.by argument 

```{r}
split_panc8 <- SplitObject(panc8, split.by = "dataset")

```
Here, the split_panc8 object consists of a list of subsetted objects which are split based on batch

```{r, echo=FALSE}
split_panc8

```

## Run seurat batch integration on 'child' projects

Multiple seurat objects can be integrated using a Chevreul function which takes in a list of seurat objects for all batches as an argument and returns an integrated seurat object containg merged batches

```{r, results="hide", cache = TRUE, eval = FALSE}
integrated_seu <- integration_workflow(split_panc8)
```

## launch app to inspect

```{r, eval = FALSE}

minimalSeuratApp(integrated_seu)

```

## view analysis details

```{r, eval = FALSE}
Misc(integrated_seu, "experiment") %>% 
  tibble::enframe() %>% 
  knitr::kable()
```
