---
title: "Visualization"
---
This article demonstrates data visualization tools present in Chevreul. Through this article we will be introducing different visualization functions, their workings, and the resulting plots produced by these functions. 

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
    # dev.args = list(png = list(type = "cairo")),
  dpi=900,
  out.width = "100%",
  message = FALSE,
  warning = FALSE
)

library(xfun)

format_table <- function(mydf){
  mydf %>% 
    kableExtra::kbl() %>% 
    kable_paper() %>% 
    scroll_box(width = "700px", height = "200px")
}

```

First set is to load Chevreul anf all other packages required 

```{r setup, }
library(chevreul)
library(Seurat)
library(tidyverse)
library(ggraph)
library(patchwork)
```

## TLDR

chevreul provides a single command to:

-   construct a Seurat object

-   filter genes by minimum expression and ubiquity

-   normalize and scale expression by any of several methods packaged in Seurat

# Run clustering on a single seurat object

By default clustering will be run at ten different resolutions between 0.2 and 2.0. Any resolution can be specified by providing the resolution argument as a numeric vector.

```{r, results=FALSE, echo=FALSE}
clustered_seu <- readRDS("clustered_seu.rds")
```

## Basic 

# Plotting features using Chevreul
 Chevreul contains a function that plots a vector of one or more features, either gene or transcript expression, overlaid on a given embedding.  This function results in a interactive, single or multi-feature plot
 
```{r}
plot_feature(clustered_seu, embedding = "umap", features = "RXRG", return_plotly = FALSE)

```

```{r}
 plot_feature(clustered_seu, embedding = "umap", features = c("RXRG", "NRL"), return_plotly = FALSE) 
```

This Chevreul function includes multiple sub-functions that ultimately helps in creation of an interactive feature plot. When plotting only one feature, an existing Seurat function, FeaturePlot() is used. This function colors single cells on a dimensional reduction plot according to a 'feature'. If multiple features are to be plotted, the joint density of all the features are plotted using a plot_density() function from the Nebulosa package. Once a feature plot of produced, this plot can be converted to an interactive plot using return_plotly = TRUE argument within the plot_feature() function

# Plotting cluster marker genes: plot_markers 

Chevreul contains a function to create a dot plot of N number of marker features that are grouped by user specified cell metadata. The features are grouped by wilcoxon rank-sum test implemented in presto.

```{r}
plot_markers(clustered_seu, metavar = "gene_snn_res.0.2", marker_method = "presto", return_plotly = FALSE)

```

In this Chevreul function, metadata variable from which to pick clusters is provided by the user through the argument metavar. From this metavar, cells are grouped based on presence of N number of marker features using wilcoxon rank sum test. In the resulting dot plot the sixe of the dot corresponds to the percentage of cells expressing the feature in each cluster and the color represents the average expression level of the feature. 

# Plot read count using plot_readcount()

The plot_readcount function in Chevreul plots a bar plot of number of read counts present in a metadata variable for each sample in the Seurat object. Here, the bars of the plot can be colored based on categorical metadata variable using the argument color.by whose default value is set to "batch"

```{r}
plot_readcount(clustered_seu, metavar = "nCount_RNA", color.by = "dataset", return_plotly = FALSE)
```

# Plot Metadata Vriable using plot_var() 

Chevreul contains a function that produces an interactive scatter plot of metadata variable, where each point in the plot represents a cell whose position on the plot is given by the cell embedding determined by the dimensional reduction technique. The default dimensionality reduction used is "umap". The group argument in plot_var() function allows the user to enter the metadata variable by which to group the cells by, its default value is set to "batch"

```{r}
plot_var(clustered_seu, group = "celltype", embedding = "umap", return_plotly = FALSE)

```
This function contains a Seurat function, DimPlot(), as sub function which produces the dimensional reduction plot.The interactive parameter, return_plotly, in plot_var when set to TRUE will convert the plot into an interactive plot using ggplotly function from R's plotly package


# plot_violin 

## intermediate

plot_all_transcripts 
plot_transcript_composition 

## complicated

cross_plot_vars 
plot_multiple_branches_heatmap 
plot_ridge 
seu_complex_heatmap 

# section about plotly

plotly is a tool to make ggplots interactive; it is only useful in an interactive context (in a shiny app)
