---
title: "Visualization"
---
This article will demonstrate different data visualization tools present in Chevreul that allows users to visualize their data. Through this article we will be introducing different visualization functions, their usage, and the resulting plots produced by these functions. 

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
    # dev.args = list(png = list(type = "cairo")),
  dpi=900,
  out.width = "100%",
  fig.width = 10,
  fig.height = 6,
  message = FALSE,
  warning = FALSE
)

library(xfun)

format_table <- function(mydf){
  mydf %>% 
    kableExtra::kbl() %>% 
    kable_paper() %>% 
    scroll_box(width = "700px", height = "200px")
}

```

First step is to load Chevreul package and all other packages required 

```{r setup, }
library(chevreul)
library(Seurat)
library(tidyverse)
library(ggraph)
library(patchwork)
```

## TLDR
### Run clustering on a single seurat object

Chevreul provides a single command, clustering_workflow(), to:

-   construct a Seurat object

-   filter genes by minimum expression and ubiquity

-   normalize and scale expression by any of several methods packaged in Seurat

This function is called Clustering Workflow, by default clustering will be run at ten different resolutions between 0.2 and 2.0. Any resolution can be specified by providing the resolution argument as a numeric vector. The new object is then saved as rds file which can be read using readRDS()

```{r, results=FALSE, echo=FALSE}
clustered_seu <- readRDS("clustered_seu.rds")
```

Chevreul contains many plotting functions that allows for visualization, these plots can be customized to display the required information.

## Plotting features 
Chevreul contains a function that plots a vector of one or more features, which is either a  gene or transcript expression, overlaid on a given embedding, resulting in an interactive, single or multi-feature plot. This Chevreul function includes multiple sub-functions that ultimately helps in creation of an interactive feature plot. 

When plotting only one feature, an existing Seurat function, FeaturePlot() is used. This function colors single cells on a dimensional reduction plot according to a 'feature'. 
 
```{r}
plot_feature(clustered_seu, embedding = "umap", features = "RXRG", return_plotly = FALSE)

```

If multiple features are to be plotted, the joint density of all the features are plotted using a plot_density() function from the R package Nebulosa embedded within the plot_feature() function.

```{r}
 plot_feature(clustered_seu, embedding = "umap", features = c("RXRG", "NRL"), return_plotly = FALSE) 
```

Once a feature plot of produced, this plot can be converted to an interactive plot using return_plotly = TRUE argument within the plot_feature() function. This function utilizes a sub-function called ggplotly() to create interactive plot that allows user to manually select cells in the plot for further investigation. 


## Plotting cluster marker genes  

Chevreul contains a function to create a dot plot of N number of marker features that are grouped by user specified cell metadata. The features are grouped by wilcoxon rank-sum test implemented in presto.

```{r}
plot_markers(clustered_seu, metavar = "gene_snn_res.0.2", marker_method = "presto")

```

In this Chevreul function, metadata variable from which to pick clusters is provided by the user through the argument metavar. From this metavar, cells are grouped based on presence of N number of marker features using wilcoxon rank sum test. In the resulting dot plot the size of the dot corresponds to the percentage of cells expressing the feature in each cluster and the color represents the average expression level of the feature. 

## Ploting read count 

The plot_readcount function in Chevreul plots a bar plot of number of reads present in a metadata variable for each sample in the Seurat object. Here, the bars of the plot can be colored based on categorical metadata variable using the argument color.by whose default value is set to "batch"

```{r}
plot_readcount(clustered_seu, metavar = "nCount_RNA", color.by = "dataset")
```

## Ploting Metadata Vriable

Chevreul contains a function that produces an interactive scatter plot of a metadata variable, where each point in the plot represents a cell whose position on the plot is given by the cell embedding determined by the dimensional reduction technique. The default dimension reduction used is "UMAP". The group argument in plot_var() function allows the user to enter the metadata variable by which to group the cells by, its default value is set to "batch".

```{r}
plot_var(clustered_seu, group = "celltype", embedding = "umap")

```
This function utilizes a Seurat function, DimPlot(), as sub function which produces the dimensional reduction plot.The interactive parameter, return_plotly, in plot_var when set to TRUE will convert the plot into an interactive plot using ggplotly function from R's plotly package


## Violin plot

To visualize the distribution of expression level of a feature in different groups of cells Chevreul draws a violin plot. This function uses Seurat's VInPlot() function as a sub-function to create the violin plot, where the metadata variable, provided to the function through the variable plot_var, is used to group the cells and based on the level of feature expression a violin plot is produced. 

```{r}
plot_violin(clustered_seu, plot_var = "celltype", features = "ISL1")
```


## Ploting all transcripts 

This function plots the expression level of all the transcripts present in a given Seurat object, coding for a feature/gene of interest. The gene of interest is specified by the argument 'features' and by default the data is superimposed on UMAP embedding. 

```{r}

plot_all_transcripts(human_gene_transcript_seu, features = "NRL") 

```


## Ploting transcrpit composition

This Chevreul function produces a visual representation of the expression levels of each transcript in a gene map. The function plot_transcript_composition() plots the proportion of reads of a given gene map to each transcript. The gene of interest is specified by the argument 'gene_symbol'. 
 
```{r, results=FALSE, echo=FALSE}

plot_transcript_composition(human_gene_transcript_seu, "NRL")

```


## Ploting ridge plots for cell cycle scoring 

The plot_ridge() function in Chevreul plots a ridge plot for cell cycle scoring of a feature, providing a visual representation of expression level of the feature in each cell cycle. This function uses two embedded Seurat functions. The function CellCycleScoring() assigns each cell a score based on its expression of G2/M and S phase markers. This data is then used by the function, RidgePlot(), which uses user provided 'features' argument to plot the ridge plot of the features.

```{r}

plot_cell_cycle_distribution(human_gene_transcript_seu, "NRL") 

plot_cell_cycle_distribution(human_gene_transcript_seu, "nFeature_RNA")

```


## Plot Complex heatmap from Seurat object 

The Chevreul function plots an annotated complex heat map from the Seurat object with a dendogram added to the left side and the top.


```{r}
top_10_features <- VariableFeatures(human_gene_transcript_seu)[1:10]

seu_complex_heatmap(human_gene_transcript_seu, features = top_10_features, group.by = "gene_snn_res.0.2")

seu_complex_heatmap(human_gene_transcript_seu, features = top_10_features, group.by = "gene_snn_res.0.2", col_arrangement = "average")

seu_complex_heatmap(human_gene_transcript_seu, features = top_10_features, group.by = c("Prep.Method", "gene_snn_res.0.2"), col_arrangement = "gene_snn_res.0.2")

seu_complex_heatmap(human_gene_transcript_seu, features = top_10_features, group.by = c("Prep.Method", "gene_snn_res.0.2"), col_arrangement = "Prep.Method")


```
