<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Tools for using seurat with single cell projects">
<title>Tools for Managing Seurat Objects • chevreul</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Tools for Managing Seurat Objects">
<meta property="og:description" content="Tools for using seurat with single cell projects">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">chevreul</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.12</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/integration.html">Integration of Seurat objects</a>
    <a class="dropdown-item" href="articles/pipeline.html">Pipeline</a>
    <a class="dropdown-item" href="articles/preprocessing.html">Preprocessing</a>
    <a class="dropdown-item" href="articles/shiny_app.html">Shiny App</a>
    <a class="dropdown-item" href="articles/visualization.html">Visualization</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cobriniklab/chevreul/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="chevreul">Chevreul<a class="anchor" aria-label="anchor" href="#chevreul"></a>
</h1></div>
<p>This package includes a set of Shiny apps for exploring single cell RNA datasets processed with <a href="https://github.com/satijalab/seurat" target="_blank" rel="noopener noreferrer" class="external-link">Seurat</a></p>
<p>A demo using a pancreas dataset from the Seurat team is available <a href="http://cobrinik-1.saban-chla.usc.edu:8080/app/seuratApp" target="_blank" rel="noopener noreferrer" class="external-link">here</a></p>
<p>There are also convenient functions for:</p>
<ul>
<li>Clustering and Dimensional Reduction of Raw Sequencing Data.</li>
<li><a href="https://satijalab.org/seurat/v3.0/pancreas_integration_label_transfer.html" target="_blank" rel="noopener noreferrer" class="external-link">Integration and Label Transfer</a></li>
<li>Louvain Clustering at a Range of Resolutions</li>
<li>Cell cycle state regression and labeling</li>
<li>RNA velocity calculation with <a href="https://velocyto.org/" target="_blank" rel="noopener noreferrer" class="external-link">Velocyto.R</a> and <a href="https://scvelo.readthedocs.io/" target="_blank" rel="noopener noreferrer" class="external-link">scvelo</a>
</li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the released version of chevreul from <a href="https://github.com/whtns/chevreul" target="_blank" rel="noopener noreferrer" class="external-link">github</a> with:</p>
<div class="section level3">
<h3 id="install-locally-and-run-in-three-steps">Install locally and run in three steps:<a class="anchor" aria-label="anchor" href="#install-locally-and-run-in-three-steps"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"whtns/chevreul"</span><span class="op">)</span></span>
<span><span class="fu">chevreul</span><span class="fu">::</span><span class="fu"><a href="reference/create_project_db.html">create_project_db</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="install-locally-custom-location-and-run-in-three-steps">Install locally (custom location!) and run in three steps:<a class="anchor" aria-label="anchor" href="#install-locally-custom-location-and-run-in-three-steps"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"whtns/chevreul"</span><span class="op">)</span></span>
<span><span class="fu">chevreul</span><span class="fu">::</span><span class="fu"><a href="reference/create_project_db.html">create_project_db</a></span><span class="op">(</span>destdir<span class="op">=</span><span class="st">'/your/path/to/app'</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cobriniklab/chevreul" class="external-link">chevreul</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tldr">TLDR<a class="anchor" aria-label="anchor" href="#tldr"></a>
</h2>
<p>chevreul provides a single command to:</p>
<ul>
<li><p>construct a Seurat object</p></li>
<li><p>filter genes by minimum expression and ubiquity</p></li>
<li><p>normalize and scale expression by any of several methods packaged in Seurat</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="run-clustering-on-a-single-seurat-object">Run clustering on a single seurat object<a class="anchor" aria-label="anchor" href="#run-clustering-on-a-single-seurat-object"></a>
</h2>
<p>By default clustering will be run at ten different resolutions between 0.2 and 2.0. Any resolution can be specified by providing the resolution argument as a numeric vector.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustered_seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/clustering_workflow.html">clustering_workflow</a></span><span class="op">(</span><span class="va">human_gene_transcript_seu</span>, experiment_name <span class="op">=</span> <span class="st">"seurat_pancreas"</span>, organism <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-a-first-look-at-a-processed-dataset-using-an-interactive-shiny-app">Get a first look at a processed dataset using an interactive shiny app<a class="anchor" aria-label="anchor" href="#get-a-first-look-at-a-processed-dataset-using-an-interactive-shiny-app"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/minimalSeuratApp.html">minimalSeuratApp</a></span><span class="op">(</span><span class="va">clustered_seu</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="set-up-a-seurat-object">Set up a seurat object<a class="anchor" aria-label="anchor" href="#set-up-a-seurat-object"></a>
</h2>
<p>We start with a gene by cell matrix of count/UMI values and a table of cell metadata</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">human_count</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">human_meta</span><span class="op">)</span></span></code></pre></div>
<p>we can create a seurat object in the usual manner</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myseu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">human_count</span>, assay <span class="op">=</span> <span class="st">"gene"</span>, meta.data <span class="op">=</span> <span class="va">human_meta</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocess-the-seurat-object">Preprocess the seurat object<a class="anchor" aria-label="anchor" href="#preprocess-the-seurat-object"></a>
</h2>
<p>Chevreul includes a handy function to preprocess the data that handles normalization and scaling required for downstream analysis. Preprocessing is performed using existing Seurat functions. If needed, parameters can be specified by the user.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myseu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/seurat_preprocess.html">seurat_preprocess</a></span><span class="op">(</span><span class="va">myseu</span><span class="op">)</span></span></code></pre></div>
<p>This single function includes sub-functions that normalizes, identifies highly variable features and scales the data:</p>
<ul>
<li>The normalizing sub-function performs log normalization using a default scale factor of 10,000.</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">preprocess_seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">myseu</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>After normalization, subset of features that exhibit high cell-to-cell variation in the dataset are identified. By default, 2,000 features per dataset are returned by this function.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">preprocess_seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">preprocess_seu</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, </span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Finally, the data is scaled by applying linear transformation. This step shifts the gene expression, so that the mean expression across cells is 0 and scales the gene expression, so that the variance across cells is 1.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pre_process_seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">preprocess_seu</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="perform-dimension-reduction">Perform dimension reduction<a class="anchor" aria-label="anchor" href="#perform-dimension-reduction"></a>
</h2>
<p>Chevreul also implements a standardized dimension reduction step to select variable features at a user-specified threshold and perform PCA, tSNE, and UMAP. The default assay the dimension reduction is being run on is “gene”.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myseu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/seurat_reduce_dimensions.html">seurat_reduce_dimensions</a></span><span class="op">(</span><span class="va">myseu</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span> </span></code></pre></div>
<p>This function includes existing seurat functions which performs dimension reduction techniques.</p>
<ul>
<li>Perform PCA: Runs a PCA dimensionality reduction.</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Dim_Red_seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">myseu</span>,features <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">myseu</span><span class="op">)</span>, </span>
<span>      do.print <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Perform tSNE: Runs t-SNE dimensionality reduction on selected features</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Dim_Red_seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunTSNE.html" class="external-link">RunTSNE</a></span><span class="op">(</span><span class="va">Dim_Red_seu</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Perform UMAP: Runs the Uniform Manifold Approximation and Projection (UMAP) dimensional reduction technique.</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Dim_Red_seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">Dim_Red_seu</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">Dim_Red_seu</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="community-detection-by-clustering">Community detection by clustering<a class="anchor" aria-label="anchor" href="#community-detection-by-clustering"></a>
</h2>
<p>Clustering analysis is performed via Louvain(default) or alternative algorithms available in Seurat. Clustering is performed at a range of resolutions with default value ranging from 0.2 to 2 and pca reduction</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/seurat_cluster.html">seurat_cluster</a></span><span class="op">(</span>seu <span class="op">=</span> <span class="va">Dim_Red_seu</span>, resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">2</span>, by <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<p>This function produces clustering analysis via two steps performed using two different sub-functions</p>
<ul>
<li><p><code>FindNeighbours</code>: This function computes the nearest neighbors for a given dataset using k-nearest neighbor algorithm.</p></li>
<li><p><code>FindClusters</code>: The output from FindNeighbours is then used to compute</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="split-included-dataset-based-on-collection-technology">Split included dataset based on collection technology<a class="anchor" aria-label="anchor" href="#split-included-dataset-based-on-collection-technology"></a>
</h2>
<p>Chevreul includes a function, <code>SplitObject</code>, which is capable of splitting the dataset into subsets based on a single attribute indicated by the split.by argument</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">split_panc8</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SplitObject.html" class="external-link">SplitObject</a></span><span class="op">(</span><span class="va">panc8</span>, split.by <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span></code></pre></div>
<p>Here, the split_panc8 object consists of a list of subsetted objects which are split based on batch</p>
</div>
<div class="section level2">
<h2 id="run-seurat-batch-integration-on-child-projects">Run seurat batch integration on ‘child’ projects<a class="anchor" aria-label="anchor" href="#run-seurat-batch-integration-on-child-projects"></a>
</h2>
<p>Multiple seurat objects can be integrated using a Chevreul function which takes in a list of seurat objects for all batches as an argument and returns an integrated seurat object containg merged batches</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">integrated_seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/integration_workflow.html">integration_workflow</a></span><span class="op">(</span><span class="va">split_panc8</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="launch-app-to-inspect">launch app to inspect<a class="anchor" aria-label="anchor" href="#launch-app-to-inspect"></a>
</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/minimalSeuratApp.html">minimalSeuratApp</a></span><span class="op">(</span><span class="va">integrated_seu</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="view-analysis-details">view analysis details<a class="anchor" aria-label="anchor" href="#view-analysis-details"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Misc.html" class="external-link">Misc</a></span><span class="op">(</span><span class="va">integrated_seu</span>, <span class="st">"experiment"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/enframe.html" class="external-link">enframe</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cobriniklab/chevreul/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/cobriniklab/chevreul/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing chevreul</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Kevin Stachelek <br><small class="roles"> Author, maintainer </small>  </li>
<li>Bhavana Bhat <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Kevin Stachelek, Bhavana Bhat.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
