<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="chevreul">
<title>Shiny App â€¢ chevreul</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Shiny App">
<meta property="og:description" content="chevreul">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">chevreul</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.12</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/integration.html">Integration of Seurat objects</a>
    <a class="dropdown-item" href="../articles/pipeline.html">Pipeline</a>
    <a class="dropdown-item" href="../articles/preprocessing.html">Preprocessing</a>
    <a class="dropdown-item" href="../articles/shiny_app.html">Shiny App</a>
    <a class="dropdown-item" href="../articles/visualization.html">Visualization</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cobriniklab/chevreul/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Shiny App</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cobriniklab/chevreul/blob/HEAD/vignettes/articles/shiny_app.Rmd" class="external-link"><code>vignettes/articles/shiny_app.Rmd</code></a></small>
      <div class="d-none name"><code>shiny_app.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cobriniklab/chevreul" class="external-link">chevreul</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="tldr">TLDR<a class="anchor" aria-label="anchor" href="#tldr"></a>
</h3>
<p>chevreul provides a single command to:</p>
<ul>
<li><p>construct a Seurat object</p></li>
<li><p>filter genes by minimum expression and ubiquity</p></li>
<li><p>normalize and scale expression by any of several methods packaged
in Seurat</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="run-clustering-on-a-single-seurat-object">Run clustering on a single seurat object<a class="anchor" aria-label="anchor" href="#run-clustering-on-a-single-seurat-object"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/minimalSeuratApp.html">minimalSeuratApp</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="app-sections">App sections<a class="anchor" aria-label="anchor" href="#app-sections"></a>
</h2>
<p>*** include text to indicate that we will be going through sections
in the app and providing info</p>
<div class="section level4">
<h4 id="integrate-projects">Integrate Projects<a class="anchor" aria-label="anchor" href="#integrate-projects"></a>
</h4>
<p>Integration of 2 or more single-cell datasets is important for its
joint analyses. Choices of cell-level and batch-level qc are highly
relevant when performing batch integration.</p>
<p>Chevreul provides tools for cell-level exploratory data analysis and
for specifying integration settings. While operating Chevreul in server
mode, separate sequencing batches can be <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8" class="external-link">integrated</a>
and compared to validate sample processing steps and exclude technical
variation in favor of relevant biological variation.</p>
<p>This section of Chevreul, includes a list of projects which can be
filtered through to select and run the integration.</p>
</div>
<div class="section level4">
<h4 id="reformat-metadata">Reformat Metadata<a class="anchor" aria-label="anchor" href="#reformat-metadata"></a>
</h4>
<p>** Make spreadsheet smaller in the app<strong> #### Arbitrary
variables can be appended to the cell metadata based on the results of
exploratory data analysis. Metadata addition can be executed by
uploading a csv with row names matching cell ids with new variables as
columns. Metadata editing can also be accomplished using a built-in
spreadsheet tool. </strong>** where is the split by option in the
app?</p>
<p>Cevreul makes it simple to subset a dataset consisting of a single
batch or batch-integrated dataset. Subsetting can be accomplished either
in a graphical setting by selection from a dimensionally reduced plot or
by specifying a properly formatted file. Subsetting of single or batch
integrated data will trigger renewal of all relevant preprocessing
steps.</p>
</div>
<div class="section level4">
<h4 id="heat-mapsviolin-plots">Heat Maps/Violin Plots<a class="anchor" aria-label="anchor" href="#heat-mapsviolin-plots"></a>
</h4>
<div class="section level5">
<h5 id="heat-maps">
<em>Heat Maps</em><a class="anchor" aria-label="anchor" href="#heat-maps"></a>
</h5>
<p>When plotting the heat map, gene/transcript expression values for
each cell are normalized by that cellâ€™s total expression then multiplied
by 10,000 and natural-log transformed before plotting.</p>
<p>Clustering model or metadata for column arrangement is taken as a
parameter to produce a complex heat map.</p>
<p>By default, 50 most highly variable genes are displayed. However, an
arbitrary lists of genes can be plotted for comparison. The
genes/transcripts are displayed in the order they are listed.</p>
<p><img src="images/Heatmap2.0.PNG"></p>
</div>
<div class="section level5">
<h5 id="violin-plots">
<em>Violin Plots</em><a class="anchor" aria-label="anchor" href="#violin-plots"></a>
</h5>
<p>Feature (gene or transcript) expression can be viewed by violin plot
based on sample metadata</p>
<p>Violin plots are a hybrid of density and box plots. It shows the
probability density of the data at different values providing a visual
representation of distribution of expression level of a feature in
different groups of cell. The horizontal line is a marker for the median
of the data and the box shows the interquartile ranges.</p>
<p>The parameters that could to be chosen here are: 1. â€˜Grouping
variableâ€™ 2. â€˜Data Type (Transformed/Raw)â€™</p>
<p><img src="images/violinPlot.PNG"></p>
</div>
</div>
<div class="section level4">
<h4 id="coverage-plots">Coverage plots<a class="anchor" aria-label="anchor" href="#coverage-plots"></a>
</h4>
<p>Coverage plots indicate the **resd depts across all transcripts for
different groups of cells within a genomic region.</p>
<p>The three user input parameters are: 1. â€˜Select a geneâ€™- Select a
gene of intreset 2. â€˜Color by variableâ€™- Select a variable by which to
group the cells by 3. â€˜Groups to displayâ€™- Select the groups to be
displayed</p>
<p><img src="images/PlotCoverage2.0.PNG" style="width:80.0%;height:80.0%"></p>
</div>
<div class="section level4">
<h4 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h4>
<p>Chevreul provides different methods of differential expression
analysis that users can choose from to determine differential expression
of gene. DE methods included in Chvreul are t-test, Wilcoxon rank-sum
test, Likelihood ratio test, and MAST</p>
<p>Running DE testing results in a data frame containing the following
information: * p_val : Unadjusted p-value * avg_log2FC : log fold-change
of the average expression between the two groups. * pct.1 : The
percentage of cells where the feature is detected in the first group *
pct.2 : The percentage of cells where the feature is detected in the
second group * p_val_adj : Adjusted p-value, based on bonferroni
correction using all features in the dataset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"Dominic.csv"</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="find-markers">Find Markers<a class="anchor" aria-label="anchor" href="#find-markers"></a>
</h4>
</div>
</div>
<div class="section level2">
<h2 id="rna-velocity">RNA Velocity<a class="anchor" aria-label="anchor" href="#rna-velocity"></a>
</h2>
<p>RNA velocity is the time-derivative of the gene expression state of a
cell. RNA velocity can be used to forecast the future state of a given
cell in a developmental or time-dependent process. seuratTools provides
visualization and processing functionality using <a href="https://velocyto.org/" class="external-link">velocyto</a>. Velocity can be represented
on UMAP or tSNE plots in multiple formats and exported as needed. RNA
velocity necessitates the calculation of spliced and unspliced
transcript counts as implemented on the command-line in velocyto. As a
result integration within seuratTools demands some familiarity with the
command-line and the availability of a properly formatted <a href="https://velocyto.org/velocyto.py/tutorial/cli.html#about-the-output-loom-file" class="external-link">loom</a>
file. This file must be generated separately from the seurat object. To
enable plotting of RNA velocity, the path to the corresponding loom file
must be provided.</p>
</div>
<div class="section level2">
<h2 id="regress-features">Regress Features<a class="anchor" aria-label="anchor" href="#regress-features"></a>
</h2>
<p>Often we would like to correct for expression variation in our
analysis due to phenomena unrelated to our focus of study. Such unwanted
effects can include cell-cycle variability or systematic differences in
mitochondrial expression.</p>
<p>Rather than excluding count values attributable to such process, we
adjust the expression of all remaining genes or transcripts in each cell
based on the sum score of relevant genes, to regress out cell-cycle
effects, for <a href="https://satijalab.org/seurat/v3.0/cell_cycle_vignette.html" class="external-link">example</a></p>
<p>Such regression can be performed for any arbitrary gene set. Some
preset gene lists are included in <code>seuratTools</code> including
cell-cycle, mitochondrial, and apoptosis related gene sets.</p>
</div>
<div class="section level2">
<h2 id="pseudotime-analysis-by-monocle-3">Pseudotime Analysis by <a href="https://cole-trapnell-lab.github.io/monocle3/" class="external-link">Monocle 3</a><a class="anchor" aria-label="anchor" href="#pseudotime-analysis-by-monocle-3"></a>
</h2>
<p>Pseudotime analysis is broken into several steps, mirroring a typical
command-line workflow. Chevreul provides graphical tools to optionally
subset an embedded dataset then calculate a trajectory graph. Root cells
can be specified and cells ordered in pseudotime with monocle3. Features
that vary significantly over pseudotime based on autocorrelation can
then be determined and grouped into modules that vary similarly over
pseudotime. Heatmap and scatter plot visualiation of pseudotime analysis
results are available at gene and module levels</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Kevin Stachelek, Bhavana Bhat.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
