<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ seuratTools</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seuratTools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cobriniklab/seuratTools">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="Monocle2_diffex.html">Monocle2_diffex()</a></code> </p>
        </td>
        <td><p>Monocle2 differential expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SetDefaultAssay.html">SetDefaultAssay()</a></code> </p>
        </td>
        <td><p>Reset the default assay of a seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_read_count_col.html">add_read_count_col()</a></code> </p>
        </td>
        <td><p>Annotate Low Read Count Category</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="allTranscripts.html">allTranscripts()</a></code> </p>
        </td>
        <td><p>Plot All Transcripts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="allTranscriptsui.html">allTranscriptsui()</a></code> </p>
        </td>
        <td><p>Plot All Transcripts UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotate_cell_cycle.html">annotate_cell_cycle()</a></code> </p>
        </td>
        <td><p>Annotate Cell Cycle</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotate_excluded.html">annotate_excluded()</a></code> </p>
        </td>
        <td><p>Annotate Exclusion Criteria</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="arrange_ptime_heatmaps.html">arrange_ptime_heatmaps()</a></code> </p>
        </td>
        <td><p>Arrange processed monocle CDS objects containing heatmaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assign_clusters_to_cds.html">assign_clusters_to_cds()</a></code> </p>
        </td>
        <td><p>Assign Clusters to CDS</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ccScore.html">ccScore()</a></code> </p>
        </td>
        <td><p>Cell Cycle Score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ccScoreui.html">ccScoreui()</a></code> </p>
        </td>
        <td><p>Cell Cycle Score UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="changeEmbedParams.html">changeEmbedParams()</a></code> </p>
        </td>
        <td><p>Change Embedding Parameters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="changeEmbedParamsui.html">changeEmbedParamsui()</a></code> </p>
        </td>
        <td><p>Change Embedding Parameters UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clustering_workflow.html">clustering_workflow()</a></code> </p>
        </td>
        <td><p>Clustering Workflow</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_human_seu_to_mouse.html">convert_human_seu_to_mouse()</a></code> </p>
        </td>
        <td><p>Convert Seurat Objects from Human to Mouse</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_mouse_seu_to_human.html">convert_mouse_seu_to_human()</a></code> </p>
        </td>
        <td><p>Convert Seurat Objects from Mouse to Human</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_seu_to_cds.html">convert_seu_to_cds()</a></code> <code><a href="convert_seu_to_cds.html">convert_seu_to_cds()</a></code> </p>
        </td>
        <td><p>Convert a Seurat Object to a Monocle Cell Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_seuv3_to_monoclev2.html">convert_seuv3_to_monoclev2()</a></code> </p>
        </td>
        <td><p>Convert a Seurat V3 object to a Monocle v2 object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_symbols_by_species.html">convert_symbols_by_species()</a></code> </p>
        </td>
        <td><p>convert gene symbols between species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_proj_list.html">create_proj_list()</a></code> </p>
        </td>
        <td><p>Create List of Projects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_proj_matrix.html">create_proj_matrix()</a></code> </p>
        </td>
        <td><p>Create a Table of single Cell Projects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cross_species_integrate.html">cross_species_integrate()</a></code> </p>
        </td>
        <td><p>Integrate Seurat Objects from Moust to Human</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="diffex.html">diffex()</a></code> </p>
        </td>
        <td><p>Differential Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="diffexui.html">diffexui()</a></code> </p>
        </td>
        <td><p>Differential Expression UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_low_rc_cells.html">filter_low_rc_cells()</a></code> </p>
        </td>
        <td><p>Filter our Cells from Seurat below read count threshold</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_merged_seu.html">filter_merged_seu()</a></code> </p>
        </td>
        <td><p>Filter a Single Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_merged_seus.html">filter_merged_seus()</a></code> </p>
        </td>
        <td><p>Filter a List of Seurat Objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findMarkers.html">findMarkers()</a></code> </p>
        </td>
        <td><p>Find Markers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findMarkersui.html">findMarkersui()</a></code> </p>
        </td>
        <td><p>Find Markers UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_all_markers.html">find_all_markers()</a></code> </p>
        </td>
        <td><p>Find All Markers at a range of resolutions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneEnrichment.html">geneEnrichment()</a></code> </p>
        </td>
        <td><p>Gene Enrichment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneEnrichmentui.html">geneEnrichmentui()</a></code> </p>
        </td>
        <td><p>Gene Enrichment UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_test_plot.html">gene_test_plot()</a></code> </p>
        </td>
        <td><p>Plot Gene expression over pseudotime on reference and query trajectories from cellAlign</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genes_to_transcripts.html">genes_to_transcripts()</a></code> </p>
        </td>
        <td><p>Ensembl Genes to Transcripts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_meta.html">get_meta()</a></code> </p>
        </td>
        <td><p>Get Metadata for a project</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_transcripts_from_seu.html">get_transcripts_from_seu()</a></code> </p>
        </td>
        <td><p>Get Transcripts in Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="global_align.html">global_align()</a></code> </p>
        </td>
        <td><p>Global Alignment of two trajectories with cellAlign</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="integrateProj.html">integrateProj()</a></code> </p>
        </td>
        <td><p>Integrate Projects Server Function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="integrateProjui.html">integrateProjui()</a></code> </p>
        </td>
        <td><p>Integrate Project UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="integration_workflow.html">integration_workflow()</a></code> </p>
        </td>
        <td><p>Integration Workflow</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="learn_graph_by_resolution.html">learn_graph_by_resolution()</a></code> <code><a href="learn_graph_by_resolution.html">learn_graph_by_resolution()</a></code> </p>
        </td>
        <td><p>Learn Monocle Graph by Resolution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="list_plot_types.html">list_plot_types()</a></code> </p>
        </td>
        <td><p>Collate list of variables to be plotted</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_counts_from_stringtie.html">load_counts_from_stringtie()</a></code> </p>
        </td>
        <td><p>Load Counts from Stringtie Files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_meta.html">load_meta()</a></code> </p>
        </td>
        <td><p>Load Project Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_seurat_from_proj.html">load_seurat_from_proj()</a></code> </p>
        </td>
        <td><p>Load Seurat Files from a signle project path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_seurat_path.html">load_seurat_path()</a></code> </p>
        </td>
        <td><p>Read in Gene and Transcript Seurat Objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="local_align.html">local_align()</a></code> </p>
        </td>
        <td><p>Align two trajectories via cellalign</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDimRed.html">plotDimRed()</a></code> </p>
        </td>
        <td><p>Plot Dimensionally Reduced Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDimRedui.html">plotDimRedui()</a></code> </p>
        </td>
        <td><p>Plot Dimensionally Reduced Data UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotReadCount.html">plotReadCount()</a></code> </p>
        </td>
        <td><p>Plot Read Count</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotReadCountui.html">plotReadCountui()</a></code> </p>
        </td>
        <td><p>Plot Read Count UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cds.html">plot_cds()</a></code> <code><a href="plot_cds.html">plot_cds()</a></code> </p>
        </td>
        <td><p>Plot a Monocle Cell Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_feature.html">plot_feature()</a></code> </p>
        </td>
        <td><p>Plot Features
This is a great function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_feature_in_ref_query_ptime.html">plot_feature_in_ref_query_ptime()</a></code> </p>
        </td>
        <td><p>Plot Expression of a Given Feature of a set of Pseudotimes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_markers.html">plot_markers()</a></code> </p>
        </td>
        <td><p>Plot Cluster Marker Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_monocle_all_ptimes.html">plot_monocle_all_ptimes()</a></code> </p>
        </td>
        <td><p>Plot a heatmap of differentially expressed genes over a query and set of reference pseudotimes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_multiple_branches_heatmap.html">plot_multiple_branches_heatmap()</a></code> </p>
        </td>
        <td><p>Plot pseudotime over multiple branches</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pseudotime.html">plot_pseudotime()</a></code> <code><a href="plot_pseudotime.html">plot_pseudotime()</a></code> </p>
        </td>
        <td><p>Plot a Monocle Cell Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pseudotime_heatmap.html">plot_pseudotime_heatmap()</a></code> <code><a href="plot_pseudotime_heatmap.html">plot_pseudotime_heatmap()</a></code> </p>
        </td>
        <td><p>Plot Pseudotime Heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_readcount.html">plot_readcount()</a></code> </p>
        </td>
        <td><p>Plot Read Count</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_ridge.html">plot_ridge()</a></code> </p>
        </td>
        <td><p>Plot Rides</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_var.html">plot_var()</a></code> </p>
        </td>
        <td><p>Plot Metadata Variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_velocity.html">plot_velocity()</a></code> </p>
        </td>
        <td><p>Plot RNA velocity Computed by Velocyto.R</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_violin.html">plot_violin()</a></code> </p>
        </td>
        <td><p>Make Violin Plots Based on Metadata and Feature Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prep_slider_values.html">prep_slider_values()</a></code> </p>
        </td>
        <td><p>Prep Slider Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="process_monocle_child.html">process_monocle_child()</a></code> </p>
        </td>
        <td><p>Preprocess a Monocle v2 object for heatmap based on provided pseudotime</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="regress_by_features.html">regress_by_features()</a></code> </p>
        </td>
        <td><p>Regress Seurat Object by Given Set of Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reintegrate_seu.html">reintegrate_seu()</a></code> </p>
        </td>
        <td><p>Reintegrate (filtered) seurat objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rename_markers.html">rename_markers()</a></code> </p>
        </td>
        <td><p>Convenience Function to transition existing seus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rename_seurat.html">rename_seurat()</a></code> </p>
        </td>
        <td><p>Give a new project name to a seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reorg_seurat_files.html">reorg_seurat_files()</a></code> </p>
        </td>
        <td><p>Reorganize seurat objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="replace_markers_in_proj.html">replace_markers_in_proj()</a></code> </p>
        </td>
        <td><p>Convenience function to replace markers in seurat objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reset_seu_meta.html">reset_seu_meta()</a></code> </p>
        </td>
        <td><p>Reset Seurat Metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rnaVelocity.html">rnaVelocity()</a></code> </p>
        </td>
        <td><p>RNA Velocity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rnaVelocityui.html">rnaVelocityui()</a></code> </p>
        </td>
        <td><p>RNA Velocity UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_BEAM.html">run_BEAM()</a></code> </p>
        </td>
        <td><p>Run BEAM from Monocle2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_enrichmentbrowser.html">run_enrichmentbrowser()</a></code> </p>
        </td>
        <td><p>Run Enrichment Browser on Differentially Expressed Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_hmap.html">run_hmap()</a></code> </p>
        </td>
        <td><p>Run heatmap from Monocle2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_rna_velocity.html">run_rna_velocity()</a></code> </p>
        </td>
        <td><p>Run RNA Velocity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_rna_velocity2.html">run_rna_velocity2()</a></code> </p>
        </td>
        <td><p>Run RNA Velocity2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_seurat_de.html">run_seurat_de()</a></code> </p>
        </td>
        <td><p>Run Seurat Differential Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="save_proj_feature_seus.html">save_proj_feature_seus()</a></code> </p>
        </td>
        <td><p>Convenience function to save all seurat objects of a given project by feature and suffix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="save_seurat.html">save_seurat()</a></code> </p>
        </td>
        <td><p>Save seurat object to &lt;project&gt;/output/sce/&lt;feature&gt;_seu.rds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scdeTest.html">scdeTest()</a></code> </p>
        </td>
        <td><p>Compare two sets of cells with scde package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_colnames_txi.html">set_colnames_txi()</a></code> </p>
        </td>
        <td><p>Set Column Names from <code>tximport</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seu_from_tibbles.html">seu_from_tibbles()</a></code> </p>
        </td>
        <td><p>Create a Seurat Object from a set of tibbles</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seu_from_tximport.html">seu_from_tximport()</a></code> </p>
        </td>
        <td><p>Create a Seurat Object from return of tximport</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratApp.html">seuratApp()</a></code> </p>
        </td>
        <td><p>Create Seurat App</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_cluster.html">seurat_cluster()</a></code> </p>
        </td>
        <td><p>Run Louvain Clustering at Multiple Resolutions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_find_markers.html">seurat_find_markers()</a></code> </p>
        </td>
        <td><p>Find Cell Type Markers in a Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_integrate.html">seurat_integrate()</a></code> </p>
        </td>
        <td><p>Batch Correct Multiple Seurat Objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_integration_pipeline.html">seurat_integration_pipeline()</a></code> </p>
        </td>
        <td><p>Run Seurat Integration</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_pipeline.html">seurat_pipeline()</a></code> </p>
        </td>
        <td><p>Run Seurat Pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_preprocess.html">seurat_preprocess()</a></code> </p>
        </td>
        <td><p>Preprocess Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_reduce_dimensions.html">seurat_reduce_dimensions()</a></code> </p>
        </td>
        <td><p>Dimensional Reduction</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seurat_reference_integrate.html">seurat_reference_integrate()</a></code> </p>
        </td>
        <td><p>Predict One Seurat Object on another merged set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stash_marker_features.html">stash_marker_features()</a></code> </p>
        </td>
        <td><p>Stash Marker Genes in a Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tableSelected.html">tableSelected()</a></code> </p>
        </td>
        <td><p>Create Table of Selected Cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tableSelectedui.html">tableSelectedui()</a></code> </p>
        </td>
        <td><p>Create Table of Selected Cells UI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="update_exp_meta.html">update_exp_meta()</a></code> </p>
        </td>
        <td><p>Update experiment Metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="update_seu_meta.html">update_seu_meta()</a></code> </p>
        </td>
        <td><p>Update Seurat Metadata from Project</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Kevin Stachelek.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


